Improve git config section validation

denik opened the pull request:
Previously we checked if ActualBranch (from .git) and Branch disagreed and issued an error unless --force was passed.

We still do that, but also:
- if --force is passed, we now log this as a warning (previously it went silent)
- we do the same logic for commit (previously unchecked)
- we do similar logic for originURL except it always logs warning and never prevent the deployment

I removed separate mutator (validate_git_details.go) and put everything into load_git_details.go.

shreyas-goenka commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -40,20 +40,12 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
+		log.Warnf(ctx, "failed to load latest commit: %s", err)
 	}
+	b.Config.Bundle.Git.Commit = commit
+	b.Config.Bundle.Git.OriginURL = repo.OriginUrl()
```
LGTM, but can you also add the `bundle: "readonly"` tag to b.Config.Bundle.Git.OriginURL? That'll remove it from the bundle schema as well.

denik replied:
thanks, done!

pietern commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -40,20 +40,12 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
+		log.Warnf(ctx, "failed to load latest commit: %s", err)
 	}
+	b.Config.Bundle.Git.Commit = commit
+	b.Config.Bundle.Git.OriginURL = repo.OriginUrl()
```
Copying what we discussed:

* It will continue to be possible to set all properties (origin, branch, commit).
* IFF we can load repo details, we warn if the origin is different, and error if branch/commit is different.

denik replied:
> It will continue to be possible to set all properties (origin, branch, commit).
> IFF we can load repo details, we warn if the origin is different, and error if branch/commit is different.

Implemented this.

denik committed:
Improve git config section validation

Previously we checked if ActualBranch (from .git) and Branch disagreed
and issued an error unless --force was passed.

We still do that, but also:
 - if --force is passed, we now log this as a warning (previously it went silent)
 - we do the same logic for commit (previously unchecked)
 - we do similar logic for originURL except it always logs warning and never
   prevent the deployment

I removed separate mutator (validate_git_details.go) and put everything into load_git_details.go.

denik commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -40,20 +40,12 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
-	if b.Config.Bundle.Git.OriginURL == "" {
-		remoteUrl := repo.OriginUrl()
-		b.Config.Bundle.Git.OriginURL = remoteUrl
+	commit, err := repo.LatestCommit()
```
Added some tests.

denik committed:
fix tests

denik committed:
tmp - debuggin issue on CI

denik committed:
restore original Inferred logic.

On CI info.CurrentBranch is "" but according to original logic it is still Inferred
because it only checks if Config value is empty string.

denik committed:
Revert "tmp - debuggin issue on CI"

This reverts commit e5a85e9e6dfb9ac2b15dd4ce409fea16d3728780.

denik committed:
clean up duplicate test case

pietern commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -33,28 +34,56 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
 		b.Config.Bundle.Git.BundleRootPath = filepath.ToSlash(relBundlePath)
 	}
+
+	config.BundleRootPath = filepath.ToSlash(relBundlePath)
```
This looks identical to the `else` branch above. Intentional?

denik replied:
No, good catch.

pietern commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -33,28 +34,56 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
+	// In case of branch and commit and absence of --force we raise severity to Error
+	diags = append(diags, checkMatch("Git branch", info.CurrentBranch, &config.Branch, b.Config.Bundle.Force)...)
+	diags = append(diags, checkMatch("Git commit", info.LatestCommit, &config.Commit, b.Config.Bundle.Force)...)
+	diags = append(diags, checkMatch("Git originURL", info.OriginURL, &config.OriginURL, true)...)
```
```suggestion
	diags = append(diags, checkMatch("Git origin URL", info.OriginURL, &config.OriginURL, true)...)
```

pietern commented on bundle/config/mutator/load_git_details.go:
```diff
@@ -33,28 +34,56 @@ func (m *loadGitDetails) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagn
+		return []diag.Diagnostic{{
+			Severity: severity,
+			Summary:  fmt.Sprintf(tmpl, field, *configValue, fetchedValue, extra),
+		}}
```
This happens only if the configuration is explicitly set. Can you include the location information here? You can access it through `b.Config.GetLocations()` for the relevant paths in the configuration.

This means the user can Cmd-click on it to jump to the problematic location.

Separate: I see this matches the existing warning and that's fine, but this should really be broken out into a short single-line summary and longer multi-line "detail" field.

andrewnester replied:
I'd also prefer to have this addressed here. The PR adds additional logic for commit and url check anyway and thus the message is different, so having location info is useful.

pietern reviewed (commented):
Thanks, this looks great!

Does the e2e coverage of `LoadGitDetails` cover this as well?

denik committed:
Update bundle/config/mutator/load_git_details.go

Co-authored-by: Pieter Noordhuis <pieter.noordhuis@databricks.com>

denik committed:
rm duplicate line

denik commented:
> Thanks, this looks great!
> 
> Does the e2e coverage of `LoadGitDetails` cover this as well?

Not, E2E, but there is https://github.com/databricks/cli/blob/main/bundle/tests/git_test.go#L45 that checks Branch validation behavior.

andrewnester reviewed (commented):
It is a potential breaking change because the validation used to happen **after** bundle variables were resolved but now it happens **before**. There's a chance (like in here https://github.com/databricks/cli/issues/1255) that someone is using variable for Git option settings and thus validation will fail
